### PR TITLE
Couple ShapeTag with Schema

### PR DESCRIPTION
**Pull request by Baccata**
Adds a constraint that enforces that all shape tags have an associated schema. This allows to completely remove the need for `SchemaIndex`, and greatly simplifies the dynamic compilation of models.

The simplification is allowed by a change to the Hints implementation which can now store two kinds of bindings : static ones and dynamic ones. Dynamic bindings are essentially a pair of ShapeId and Document. Because ShapeTags now have an associated schema, this schema can be used to deserialise the Document on the fly, which means that the Dynamic compiler does not need to receive schemas in advance anymore.

The cost is that some manually defined shapes need to receive a Schema now.

### CODE REVIEW

**Review by daddykotex — APPROVED** [review]
I love what I see, lots of red!

The only downside is that we have to trust you with the handcrafted schema :P

File: modules/codegen/src/smithy4s/codegen/Renderer.scala [hunk 1]
```diff
@@ -315,14 +315,14 @@ private[codegen] class Renderer(compilationUnit: CompilationUnit) { self =>
   private def renderProtocol(name: String, hints: List[Hint]): Lines = {
     hints.collectFirst({ case p: Hint.Protocol => p }).foldMap { protocol =>
       val protocolTraits = protocol.traits
-        .map(t => line"${t.namespace}.${t.name.capitalize}")
```
[the remark this answers is not in the document]
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/Renderer.scala:318]: I ... did indeed 😄

[the remark this answers is not in the document]
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/Renderer.scala:318]: (for context, the capitalize call was there to create a class instance. We're not creating a class instance anymore)

**Review by kubukoz — APPROVED** [review]
playground is happy, really good to see this!

**Pull request merged**